cpp/src/transform/encode.cu

```diff
@@ -0,0 +1,83 @@
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/search.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <numeric>
+
+namespace cudf {
+namespace detail {
+
+std::pair<std::unique_ptr<table>, std::unique_ptr<column>> encode(
+  table_view const& input_table, rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  std::vector<size_type> columns(input_table.num_columns());
+  std::iota(columns.begin(), columns.end(), 0);
+
+  // side effects of this function we are now dependent on:
+  // - resulting column elements are sorted ascending
+  // - nulls are sorted to the beginning
+  auto keys_table = cudf::detail::drop_duplicates(
+    input_table, columns, duplicate_keep_option::KEEP_FIRST, null_equality::EQUAL, mr, stream);
+
+  if (cudf::has_nulls(keys_table->view())) {
+    // Rows with nulls appear at the top of `keys_table`, but we want them to appear at
+    // the bottom. Below, we rearrange the rows so that nulls appear at the bottom:
```

Comment:
rather than compensate for `drop_duplicates` null placement behavior we could update `drop_duplicates` to include support for `null_order`. If not in this PR, then as part of a tech debt issue?

Reply:
Yeah -- that would be ideal. Maybe it doesn't have to be exposed in the public `drop_duplicates`, but `detail::drop_duplicates` could support it. Something to discuss in a separate issue, but can add a TODO here to update when that is resolved.